### PR TITLE
docs(ref): Establish publish best practices

### PR DESCRIPTION
**Pull request by epage**
Inspired by #841 and conversations at RustConf.

At RustConf, someone was talking to me about the lack or inconsistent use of best practices within the crate ecosystem.
We have a challenge of needing to help users get up and going who can't research every last decision while not over-encouraging the status quo, leading to stagnation.

The list of tools included was based on a mixture of
- Download counts
- Word of mouth exposure
- Unique aspects covered
- Maturity

I do have a bias in this conversation as the maintainer of one of these tools which is why I decided to pick an ordering without judgement. I picked alphabetical ordering before sorting and realizing this puts my own crate first.

### CODE REVIEW

**Review by hi-rustin — APPROVED** [review]
Looks good to me. Thanks!

File: src/doc/src/reference/publishing.md [hunk 1]
```diff
@@ -136,6 +136,19 @@ In order to release a new version, change [the `version` value](manifest.md#the-
...
+> Each version should include:
+> - A changelog entry, preferrably [manually curated](https://keepachangelog.com/en/1.0.0/) though a generated one is better than nothing
+> - A [git tag](https://git-scm.com/book/en/v2/Git-Basics-Tagging) pointing to the published commit
+>
+> Examples of third-party tools that are representative of different workflows include (in alphabetical order):
+> - [cargo-release](https://crates.io/crates/cargo-release)
+> - [cargo-smart-release](https://crates.io/crates/cargo-smart-release)
+> - [release-plz](https://crates.io/crates/release-plz)
+>
+> For more, see [crates.io](https://crates.io/search?q=cargo%20release)
```
Comment (weihanglo) [inline src/doc/src/reference/publishing.md:150]:
```suggestion
> For more, see [crates.io](https://crates.io/search?q=cargo%20release).
```

Comment (weihanglo) [reply src/doc/src/reference/publishing.md:150]: I am not sure if "cargo release" is the correct keyword, but yet I can't see any better alternative 😆.

**Review by weihanglo — COMMENTED** [review]
What's the criteria if somebody wants to put their release tool in the list? Should we make it clear in this PR thread, or we'll discuss it case by case?

Apart from that, looks nice!

File: src/doc/src/reference/publishing.md (hunk 1 above)
Comment (epage) [reply src/doc/src/reference/publishing.md]: Typo fixed.

Unsure how to make the search term any better. It seems to cover a lot of cases for command-line focused releases though that leaves out tools like release-plz.  The `release` keyword isn't universally used and is too vague, covering a lot of other cases.

**Comment by epage** [comment]
> What's the criteria if somebody wants to put their release tool in the list? Should we make it clear in this PR thread, or we'll discuss it case by case?

I've noted in the PR why I chose the crates I did.  I think we'll need to handle it on a case-by-case basis though I expect a lot of cases to be obvious (brand-new tool that barely does anything is likely not something worth including).

**Review by weihanglo — APPROVED** [review]

**Comment by weihanglo** [comment]
Thanks!

 @bors r+

**Pull request merged**